xds/src/main/java/io/grpc/xds/internal/TlsXdsCredentialsProvider.java

```diff
@@ -16,21 +16,101 @@
 
 package io.grpc.xds.internal;
 
+import com.google.protobuf.Duration;
+import com.google.protobuf.util.Durations;
 import io.grpc.ChannelCredentials;
 import io.grpc.TlsChannelCredentials;
+import io.grpc.internal.GrpcUtil;
+import io.grpc.internal.JsonUtil;
+import io.grpc.util.AdvancedTlsX509KeyManager;
+import io.grpc.util.AdvancedTlsX509TrustManager;
 import io.grpc.xds.XdsCredentialsProvider;
+import java.io.File;
+import java.text.ParseException;
 import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
 
 /**
  * A wrapper class that supports {@link TlsChannelCredentials} for Xds
  * by implementing {@link XdsCredentialsProvider}.
  */
 public final class TlsXdsCredentialsProvider extends XdsCredentialsProvider {
+  private static final Logger logger = Logger.getLogger(TlsXdsCredentialsProvider.class.getName());
   private static final String CREDS_NAME = "tls";
+  private static final String CERT_FILE_KEY = "certificate_file";
+  private static final String KEY_FILE_KEY = "private_key_file";
+  private static final String ROOT_FILE_KEY = "ca_certificate_file";
+  private static final String REFRESH_INTERVAL_KEY = "refresh_interval";
+  private static final long REFRESH_INTERVAL_DEFAULT = 600L;
+  private static final ScheduledExecutorServiceFactory scheduledExecutorServiceFactory =
+      ScheduledExecutorServiceFactory.DEFAULT_INSTANCE;
 
   @Override
   protected ChannelCredentials newChannelCredentials(Map<String, ?> jsonConfig) {
-    return TlsChannelCredentials.create();
+    TlsChannelCredentials.Builder builder = TlsChannelCredentials.newBuilder();
+
+    if (jsonConfig == null) {
+      return builder.build();
+    }
+
+    // use refresh interval from bootstrap config if provided; else defaults to 600s
+    long refreshIntervalSeconds = REFRESH_INTERVAL_DEFAULT;
+    String refreshIntervalFromConfig = JsonUtil.getString(jsonConfig, REFRESH_INTERVAL_KEY);
+    if (refreshIntervalFromConfig != null) {
+      try {
+        Duration duration = Durations.parse(refreshIntervalFromConfig);
+        refreshIntervalSeconds = Durations.toSeconds(duration);
+      } catch (ParseException e) {
+        logger.log(Level.WARNING, "Unable to parse refresh interval", e);
+        return null;
+      }
+    }
+
+    // use trust certificate file path from bootstrap config if provided; else use system default
+    String rootCertPath = JsonUtil.getString(jsonConfig, ROOT_FILE_KEY);
+    if (rootCertPath != null) {
+      try {
+        AdvancedTlsX509TrustManager trustManager = AdvancedTlsX509TrustManager.newBuilder().build();
+        trustManager.updateTrustCredentials(
+            new File(rootCertPath),
+            refreshIntervalSeconds,
+            TimeUnit.SECONDS,
+            scheduledExecutorServiceFactory.create());
```

Comment:
We should be shutting down the executors as well, when no longer needed. We also shouldn't create two executors here; we can trivially share between the key and trust managers.